README.txt
LAStools - award-winning software for efficient LiDAR processing (with LASzip)

You find the individual LAStools in the .\lastools\bin directory. Start
them by double-clicking or run them in the DOS command line. The ArcGIS
toolbox can be found in the .\lastools\ArcGIS_toolbox directory.

compant page:    http://rapidlasso.com
latest updates:  http://lastools.org
user group:      http://groups.google.com/group/lastools
twitter feed:    http://twitter.com/lastools
facebook page:   http://facebook.com/lastools
linkedin group:  http://linkedin.com/groups?gid=4408378

* lastool.exe is a combined GUI for multiple LAStools (each tool has its own GUI)

* lasground.exe extracts the bare-earth by classifying all ground points
* lasoverlap.exe checks overlap & vertical/horizontal alignment of flight lines
* lascontrol.exe quality checks elevations for a list of control points 
* lasclassify.exe finds buildings and the vegetation above the ground
* lascolor.exe colors the LAS points based on ortho imagery in TIF format 
* lasgrid.exe grids onto min/max/avg/std elevation, intensity, or density rasters
* lascanopy.exe computes basic raster metrics for forestry applications
* lasboundary.exe extracts a boundary polygon that encloses the points
* lasheight.exe computes for each point its height above the ground
* lasclip.exe clips points against building footprints / swath boundaries
* lastile.exe tiles huge amounts of LAS points into square tiles
* laszip.exe compresses the LAS files in a completely lossless manner
* lasinfo.exe prints out a quick overview of the contents of a LAS file
* lasindex.exe creates a spatial index LAX file for fast spatial queries
* txt2las.exe converts LIDAR data from ASCII text to binary LAS format
* las2txt.exe turns LAS into human-readable and easy-to-parse ASCII
* lasmerge.exe merges several LAS files into one
* lassplit.exe splits points of one (or many) LAS files into flight lines
* lassort.exe sorts points by gps_time, point_source, or into spatial proximity
* las2las.exe extracts last returns, clips, subsamples, translates, etc ...
* lasduplicate.exe removes duplicate points (with identical x and y, z optional) 
* lasthin.exe thins lowest / highest / random LAS points via a grid
* las2tin.exe triangulates the points of a LAS file into a TIN
* las2dem.exe rasters (via a TIN) into elevation/slope/intensity/rgb DEMs
* las2iso.exe extracts, optionally simplified, elevation contours
* lasview.exe visualizes a LAS file with a simple OpenGL viewer
* lasprecision.exe analyses the actual precision of the LIDAR points
* las2shp.exe turns binary LAS into ESRI's Shapefile format
* shp2las.exe turns an ESRI's Shapefile into binary LAS

* blast2dem.exe rasters like las2dem but with streaming TINs for billions of points. 

For Windows all binaries are included. They can also be compiled from the
source code and you can find the MSVC6.0 project files there as well. For
Linux the makefiles and many sources are included. Simply go into the root
directory and run 'make':

unzip lastools.zip
cd lastools/
make

The compiled binary executables are or will be in the ./lastools/bin directory.

---

Please read the "LICENSE.txt" file for information on the legal use and licensing
of LAStools. I would also really like it if you would send me an email and tell me
what you use LAStools for and what features and improvements you could need. 

(c) 2007-2014 martin.isenburg@rapidlasso.com - http://rapidlasso.com


LASlib/CHANGES.txt
9 July 2014 -- LASlib: make LASreader_txt parse piped ASCII text with '-stdin' and '-itxt'
9 July 2014 -- LASlib: new transforms (-set_withheld_flag 0 -set_synthetic_flag 1 -set_keypoint_flag 0)
22 March 2014 -- LASlib: allow more flavours of ASC grids for on-the-fly conversion to LAS
21 February 2014 -- LASlib: renaming of clip filters (-clip_z to -keep_z, -clip_z_between to -drop_z) 
7 February 2014 -- LASlib: renaming of several variables (xyz -> XYZ, ...) for github launch
30 December 2013 -- lasinfo: produce histogram for attributes in extra bytes with '-histo 0 5' 
9 December 2013 -- LASlib: bug fix and improved writing of new LAS 1.4 point types
25 November 2013 -- LASlib: new capability to filter points based on RGB band ranges  
10 October 2013 -- LASlib: can now read FUSION DTM rasters by on-the-fly conversion to LAS
3 October 2013 -- LASlib: fix for reading and merging on-the-fly converted rasters (e.g. BIL, ASC)
30 September 2013 -- txt2las: bug fixed for adding EXTRA_BYTES with scales and offsets
22 September 2013 -- LASlib: input files paths can now contain both '/' and '\'
18 September 2013 -- LASlib: output WARNING if -odir "c:\" or -odir "..\lidar\" is used (windows)
29 July 2013 -- lasinfo: fixed output directory bug when running with multiple cores
6 May 2013 -- lasinfo: improved handling of LAS 1.4 files
1 May 2013 -- fixed LAZ reading bug (50000 points only) for download-truncated LAZ files
9 April 2013 -- LASlib: output WARNING if -odir "c:\" or -odir "..\lidar\" is used (linux)
15 March 2013 -- corrected interplay of '-ocut 2' and '-odix _g' when used at the same time
28 December 2012 -- added '-clip_z_between 12 14.51' and '-clip_raw_z_between 1200 1451'
30 September 2012 -- extended classification for new point types is written correctly
13 September 2012 -- improved lasinfo output and fixed laszip bug for LAS 1.4 point types
22 August 2012 -- added the '-pipe_on' option for a multi-stage LAStools pipeline
11 August 2012 -- added on-the-fly buffered reading of LiDAR files (efficient with LAX)
9 July 2012 -- fixed crash that occured when input had a corrupt VLRs without payload
8 April 2012 -- added LASreader_bil for native read support of binary BIL grids
27 March 2012 -- added LASreader_asc for native read support of ESRI ASC grids
26 March 2012 -- even better output file naming with '-ocut 5' 
19 March 2012 -- better output naming (-odix/-odir) + overwrite check for point producing tools
14 February 2012 -- specify output directory with '-odir' and file name appedix '-odix'
16 January 2012 -- basic LAS 1.4 write support (down-converted points are up-converted)
28 December 2011 -- (limited) read support for LAS 1.4 (new point types are down-converted)
16 December 2011 -- support for reading ESRI' SHP format for types 1,11,21,8,18,28 
9 December 2011 -- laszip: added GUI (if called without arguments)
1 December 2011 -- laszip: fixed seek bug for interrupted LAZ files (e.g. ftp, CTRL^C)
26 October 2011 -- laszip: prototype for LAS 1.3 full waveform compression
26 October 2011 -- read / write interface for full waveform LAS 1.3 data
6 October 2011 -- large file support (fseek64) for spatial indexing and compression
6 October 2011 -- laszip: ability to read LAZ with missing chunk table
5 September 2011 -- support for reading and writing Terrasolid's BIN format
23 June 2011 -- turned on LASzip version 2.0 compressor with chunking 
11 June 2011 -- support for over 4,294,967,295 points in merged LAS/LAZ or for ASCII
31 May 2011 -- new spatial data base functionality for folders of LAS/LAZ files
21 May 2011 -- most tools can now batch process with wildcards or multiple file names
17 May 2011 -- laszip: enabling batch processing 
15 May 2011 -- laszip: improved GPS11v2 compression for interleaved flight-lines
13 May 2011 -- all tools can now read spatially indexed data if LAX file
13 May 2011 -- moved indexing, filtering, transforming into LASreader
9 May 2011 -- laszip: the chunked compressor now allows variable chunk sizes
27 April 2011 -- the LASreader of the LASlib interface now allows to seek()
25 April 2011 -- laszip: chunked compression for random access decompression
22 April 2011 -- lastransform: many more ways to transform 
11 April 2011 -- added lasreader_txt and laswriter_txt for on-the-fly conversion
20 March 2011 -- laszip: added POINT10 RGB, GPS, BYTE compressor versions 2 
19 March 2011 -- lasfilter: added various simple thinning filters
18 March 2011 -- all: added null output -nil and compressed input and output -v2
28 January 2011 -- all: major clean up for LASlib (with LASzip) re-release
24 January 2011 -- lasreader: introduced LASreadOpener
21 January 2011 -- lasreader: turned into abstract reader to support multiple files
21 January 2011 -- lastransform: (re-)created after matt told me about the optech dashmap bug
20 January 2011 -- lasreadermerged: created missing Livermore and my Extra Virgin Olive Oil
10 January 2011 -- all: licensing change for LGPL release and liblas integration
7 January 2011 -- read/writeitemraw: introduced swap buffers to reduce number of fwrite calls
25 December 2010 -- lasfilter & lasutility: created after swinging in Mara's hammock for hours
16 December 2010 -- lasdefinitions: updated to support generic LASitem point formats
12 December 2010 -- bytestream: refactored after Howard got pushy (-;
10 December 2010 -- integercompressor: unified for all entropy coders at Baeckerei Schaefer
8 December 2010 -- arithmetic & entropy: unified framework for all entropy coders
7 December 2010 -- read & write in items after getting invited to KAUST in Saudi Arabia
3 December 2010 -- all: updated to (somewhat) support LAS format 1.3
30 October 2009 -- arithmetic: refactoring Amir Said's FastAC code
31 October 2009 -- integercompressor: switched from the Rangecoder to the Entropycoder
7 September 2008 -- all: updated to support LAS format 1.2 
11 June 2007 -- lasdefinitions: number of return / scan direction bitfield order was wrong

LASlib/inc/lasreader.hpp
/*
===============================================================================

  FILE:  lasreader.hpp
  
  CONTENTS:
  
    Interface to read LIDAR points from the LAS format versions 1.0 - 1.3 and
    per on-the-fly conversion from simple ASCII files.

  PROGRAMMERS:

    martin.isenburg@rapidlasso.com  -  http://rapidlasso.com

  COPYRIGHT:

    (c) 2007-2014, martin isenburg, rapidlasso - fast tools to catch reality

    This is free software; you can redistribute and/or modify it under the
    terms of the GNU Lesser General Licence as published by the Free Software
    Foundation. See the LICENSE.txt file for more information.

    This software is distributed WITHOUT ANY WARRANTY and without even the
    implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
  
  CHANGE HISTORY:
  
     7 February 2014 -- added option '-apply_file_source_ID' when reading LAS/LAZ
    22 August 2012 -- added the '-pipe_on' option for a multi-stage LAStools pipeline
    11 August 2012 -- added on-the-fly buffered reading of LiDAR files (efficient with LAX)
    5 September 2011 -- support for reading Terrasolid's BIN format
    11 June 2011 -- billion point support: p_count & npoints are 64 bit counters
    9 April 2011 -- added capability to read on-the-fly conversion from ASCII
    24 January 2011 -- introduced LASreadOpener
    21 January 2011 -- turned into abstract reader to support multiple files
    3 December 2010 -- updated to (somewhat) support LAS format 1.3
    7 September 2008 -- updated to support LAS format 1.2 
    18 February 2007 -- created after repairing 2 vacuum cleaners in the garden
  
===============================================================================
*/
#ifndef LAS_READER_HPP
#define LAS_READER_HPP

#include "lasdefinitions.hpp"

class LASindex;
class LASfilter;
class LAStransform;
class ByteStreamIn;

class LASreader
{
public:
  int my_count;
  LASheader header;
  LASpoint point;

  I64 npoints;
  I64 p_count;

  virtual I32 get_format() const = 0;
  virtual BOOL has_layers() const { return FALSE; };

  void set_index(LASindex* index);
  inline LASindex* get_index() const { return index; };
  virtual void set_filter(LASfilter* filter);
  inline LASfilter* get_filter() const { return filter; };
  virtual void set_transform(LAStransform* transform);
  inline LAStransform* get_transform() const { return transform; };

  inline U32 get_inside() const { return inside; };
  virtual BOOL inside_none();
  virtual BOOL inside_tile(const F32 ll_x, const F32 ll_y, const F32 size);
  inline F32 get_t_ll_x() const { return t_ll_x; };
  inline F32 get_t_ll_y() const { return t_ll_y; };
  inline F32 get_t_size() const { return t_size; };
  virtual BOOL inside_circle(const F64 center_x, const F64 center_y, const F64 radius);
  inline F64 get_c_center_x() const { return c_center_x; };
  inline F64 get_c_center_y() const { return c_center_y; };
  inline F64 get_c_radius() const { return c_radius; };
  virtual BOOL inside_rectangle(const F64 min_x, const F64 min_y, const F64 max_x, const F64 max_y);
  inline F64 get_r_min_x() const { return r_min_x; };
  inline F64 get_r_min_y() const { return r_min_y; };
  inline F64 get_r_max_x() const { return r_max_x; };
  inline F64 get_r_max_y() const { return r_max_y; };

  virtual BOOL seek(const I64 p_index) = 0;
  BOOL read_point() { return (this->*read_simple)(); };

  inline void compute_coordinates() { point.compute_coordinates(); };

  inline F64 get_min_x() const { return header.min_x; };
  inline F64 get_min_y() const { return header.min_y; };
  inline F64 get_min_z() const { return header.min_z; };

  inline F64 get_max_x() const { return header.max_x; };
  inline F64 get_max_y() const { return header.max_y; };
  inline F64 get_max_z() const { return header.max_z; };

  inline F64 get_x() const { return get_x(point.get_X()); };
  inline F64 get_y() const { return get_y(point.get_Y()); };
  inline F64 get_z() const { return get_z(point.get_Z()); };

  inline F64 get_x(const I32 x) const { return header.get_x(x); };
  inline F64 get_y(const I32 y) const { return header.get_y(y); };
  inline F64 get_z(const I32 z) const { return header.get_z(z); };

  inline I32 get_X(const F64 x) const { return header.get_X(x); };
  inline I32 get_Y(const F64 y) const { return header.get_Y(y); };
  inline I32 get_Z(const F64 z) const { return header.get_Z(z); };

  virtual ByteStreamIn* get_stream() const = 0;
  virtual void close(BOOL close_stream=TRUE) = 0;

  LASreader();
  virtual ~LASreader();

protected:
  virtual BOOL read_point_default() = 0;

  LASindex* index;
  LASfilter* filter;
  LAStransform* transform;

  U32 inside;
  F32 t_ll_x, t_ll_y, t_size, t_ur_x, t_ur_y;
  F64 c_center_x, c_center_y, c_radius, c_radius_squared;
  F64 r_min_x, r_min_y, r_max_x, r_max_y;
  F64 orig_min_x, orig_min_y, orig_max_x, orig_max_y;

private:
  BOOL (LASreader::*read_simple)();
  BOOL (LASreader::*read_complex)();

  BOOL read_point_filtered();
  BOOL read_point_transformed();
  BOOL read_point_filtered_and_transformed();

  BOOL read_point_inside_tile();
  BOOL read_point_inside_tile_indexed();
  BOOL read_point_inside_circle();
  BOOL read_point_inside_circle_indexed();
  BOOL read_point_inside_rectangle();
  BOOL read_point_inside_rectangle_indexed();
};

#include "laswaveform13reader.hpp"

class LASreadOpener
{
public:
  void set_io_ibuffer_size(I32 io_ibuffer_size);
  inline I32 get_io_ibuffer_size() const { return io_ibuffer_size; };
  const char* get_file_name() const;
  void set_file_name(const char* file_name, BOOL unique=FALSE);
  BOOL add_file_name(const char* file_name, BOOL unique=FALSE);
  BOOL add_list_of_files(const char* list_of_files, BOOL unique=FALSE);
  void delete_file_name(U32 file_name_id);
  BOOL set_file_name_current(U32 file_name_id);
  U32 get_file_name_number() const;
  const char* get_file_name(U32 number) const;
  I32 get_file_format(U32 number) const;
  void set_merged(const BOOL merged);
  BOOL is_merged() const { return merged; };
  void set_buffer_size(const F32 buffer_size);
  F32 get_buffer_size() const;
  void set_neighbor_file_name(const char* neighbor_file_name, BOOL unique=FALSE);
  BOOL add_neighbor_file_name(const char* neighbor_file_name, BOOL unique=FALSE);
  void set_auto_reoffset(const BOOL auto_reoffset);
  inline BOOL is_auto_reoffset() const { return auto_reoffset; };
  void set_files_are_flightlines(const BOOL files_are_flightlines);
  inline BOOL are_files_flightlines() const { return files_are_flightlines; };
  void set_apply_file_source_ID(const BOOL apply_file_source_ID);
  inline BOOL applying_file_source_ID() const { return apply_file_source_ID; };
  void set_scale_factor(const F64* scale_factor);
  inline const F64* get_scale_factor() const { return scale_factor; };
  void set_offset(const F64* offset);
  inline const F64* get_offset() const { return offset; };
  void set_translate_intensity(F32 translate_intensity);
  void set_scale_intensity(F32 scale_intensity);
  void set_translate_scan_angle(F32 translate_scan_angle);
  void set_scale_scan_angle(F32 scale_scan_angle);
  void add_attribute(I32 data_type, const char* name, const char* description=0, F64 scale=1.0, F64 offset=0.0);
  void set_parse_string(const char* parse_string);
  void set_skip_lines(I32 skip_lines);
  void set_populate_header(BOOL populate_header);
  void set_pipe_on(BOOL pipe_on);
  const char* get_parse_string() const;
  void usage() const;
  void set_inside_tile(const F32 ll_x, const F32 ll_y, const F32 size);
  void set_inside_circle(const F64 center_x, const F64 center_y, const F64 radius);
  void set_inside_rectangle(const F64 min_x, const F64 min_y, const F64 max_x, const F64 max_y);
  BOOL parse(int argc, char* argv[]);
  BOOL is_piped() const;
  BOOL is_buffered() const;
  BOOL is_header_populated() const;
  BOOL active() const;
  BOOL is_inside() const;
  I32 unparse_inside(CHAR* string) const;
  void set_filter(LASfilter* filter);
  const LASfilter* get_filter() { return filter; };
  void set_transform(LAStransform* transform);
  const LAStransform* get_transform() { return transform; };
  void reset();
  LASreader* open(const char* other_file_name=0);
  BOOL reopen(LASreader* lasreader, BOOL remain_buffered=TRUE);
  LASwaveform13reader* open_waveform13(const LASheader* lasheader);
  LASreadOpener();
  ~LASreadOpener();

  int
  getIsMpi () const
  {
    return is_mpi;
  }

  void
  setIsMpi (int isMpi)
  {
    is_mpi = isMpi;
  }

private:
#ifdef _WIN32
  BOOL add_file_name_single(const char* file_name, BOOL unique=FALSE);
  BOOL add_neighbor_file_name_single(const char* neighbor_file_name, BOOL unique=FALSE);
#endif
  int is_mpi;
  I32 io_ibuffer_size;
  char** file_names;
  const char* file_name;
  BOOL merged;
  U32 file_name_number;
  U32 file_name_allocated;
  U32 file_name_current;
  F32 buffer_size;
  char** neighbor_file_names;
  U32 neighbor_file_name_number;
  U32 neighbor_file_name_allocated;
  BOOL comma_not_point;
  F64* scale_factor;
  F64* offset;
  BOOL auto_reoffset;
  BOOL files_are_flightlines;
  BOOL apply_file_source_ID;
  BOOL itxt;
  BOOL ipts;
  BOOL iptx;
  F32 translate_intensity;
  F32 scale_intensity;
  F32 translate_scan_angle;
  F32 scale_scan_angle;
  I32 number_attributes;
  I32 attribute_data_types[10];
  char* attribute_names[10];
  char* attribute_descriptions[10];
  F64 attribute_scales[10];
  F64 attribute_offsets[10];
  char* parse_string;
  I32 skip_lines;
  BOOL populate_header;
  BOOL pipe_on;
  BOOL use_stdin;
  BOOL unique;

  // optional extras
  LASindex* index;
  LASfilter* filter;
  LAStransform* transform;

  // optional area-of-interest query (spatially indexed) 
  F32* inside_tile;
  F64* inside_circle;
  F64* inside_rectangle;
};

#endif
